storage/pebble/registers_pruner.go

```diff
@@ -0,0 +1,326 @@
+	// TODO: Clarify whether batching can be used here for getting the first and latest heights,
+	//       and for updating the first height.
+	firstHeight, err := firstStoredHeight(p.db)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to get first height from register storage: %w", err))
+	}
+
+	latestHeight, err := latestStoredHeight(p.db)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to get latest height from register storage: %w", err))
+	}
+
+	pruneHeight := latestHeight - p.pruneThreshold
+
+	if pruneHeight-firstHeight > p.pruneInterval {
+		p.logger.Info().Uint64("prune_height", pruneHeight).Msg("pruning storage")
+
+		// update first indexed height
+		err = p.updateFirstStoredHeight(pruneHeight)
+		if err != nil {
+			ctx.Throw(fmt.Errorf("failed to update first height for register storage: %w", err))
+		}
+
+		err := p.pruneUpToHeight(pruneHeight)
+		if err != nil {
+			ctx.Throw(fmt.Errorf("failed to prune: %w", err))
+		}
+
+		if p.metrics != nil {
+			p.metrics.NumberOfBlocksPruned(pruneHeight - firstHeight)
+		}
+	}
+}
+
+// pruneUpToHeight prunes all entries in the database with heights less than or equal
+// to the specified pruneHeight. For each register prefix, it keeps the earliest entry
+// that has a height less than or equal to pruneHeight, and deletes all other entries
+// with lower heights.
+//
+// This function iterates over the database keys, identifies keys to delete in batches,
+// and uses the batchDelete function to remove them efficiently.
+//
+// Parameters:
+//   - pruneHeight: The maximum height of entries to prune.
+//
+// No errors are expected during normal operations.
+func (p *RegisterPruner) pruneUpToHeight(pruneHeight uint64) error {
+	p.setupThrottleDelay()
+	defer p.throttleTicker.Stop()
+
+	prefix := []byte{codeRegister}
+	itemsPerBatch := deleteItemsPerBatch
+	var batchKeysToRemove [][]byte
+
+	start := time.Now()
+
+	err := func(r pebble.Reader) error {
+		options := pebble.IterOptions{
+			LowerBound: prefix,
+			UpperBound: []byte{codeFirstBlockHeight},
+		}
+
+		it, err := r.NewIter(&options)
+		if err != nil {
+			return fmt.Errorf("can not create iterator: %w", err)
+		}
+		defer it.Close()
+
+		var lastRegisterID flow.RegisterID
+		var keepKeyFound bool
+
+		for it.SeekGE(prefix); it.Valid(); it.Next() {
+			key := it.Key()
+
+			keyHeight, registerID, err := lookupKeyToRegisterID(key)
+			if err != nil {
+				return fmt.Errorf("malformed lookup key %v: %w", key, err)
+			}
+
+			if p.metrics != nil {
+				p.metrics.ElementVisited()
+			}
+
+			// New register prefix, reset the state
+			if !keepKeyFound || lastRegisterID != registerID {
+				keepKeyFound = false
+				lastRegisterID = registerID
+			}
+
+			if keyHeight <= pruneHeight {
+				if !keepKeyFound {
+					// Keep the first entry found for this registerID that is <= pruneHeight
+					keepKeyFound = true
+					continue
+				}
+
+				// Otherwise, mark this key for removal
+				// Create a copy of the key to avoid memory issues
```

Comment:
curious, is this copy necessary? I'm not sure myself.

Reply:
The copy is necessary because the iterator reuses the memory for each key. Without the copy, the key could be modified in subsequent iterations, leading to incorrect data being added to `batchKeysToRemove`.

We also use the copy approach in `Registers` with the iterator's `value`.
https://github.com/onflow/flow-go/blob/b93c60f2a3dff31d8f76711822d4873cdce4a0eb/storage/pebble/registers.go#L87-L89